modules/widgets/instagram.php

```diff
@@ -0,0 +1,497 @@
+		// https://wpcom.trac.automattic.com/browser/trunk/wp-content/rest-api-plugins/endpoints/sites-instagram.php?rev=204654#L88
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			$token = Keyring::init()->get_token_store()->get_token( array( 'type' => 'access', 'id' => $token_id ) );
+			if ( get_current_user_id() !== (int) $token->meta['user_id'] ) {
+				return wp_send_json_error( array( 'message' => 'not_authorized' ), 403 );
+			}
+		}
+
+		$this->update_widget_token_id( $token_id, $widget_id );
+		$this->update_widget_token_legacy_status( false );
+
+		return wp_send_json_success( null, 200 );
+	}
+
+	/**
+	 * Updates the widget's option in the database to show if it is for legacy API or not.
+	 *
+	 * @param bool $is_legacy_token A flag to indicate if a token is for the legacy Instagram API.
+	 */
+	public function update_widget_token_legacy_status( $is_legacy_token ) {
+		$widget_options = $this->get_settings();
+
+		if ( ! is_array( $widget_options[ $this->number ] ) )
+			$widget_options[ $this->number ] = $this->defaults;
+
+		$widget_options[ $this->number ]['is_legacy_token'] = $is_legacy_token;
+		$this->save_settings( $widget_options );
+
+		return $is_legacy_token;
+	}
+
+	private function get_token_status( $token_id ) {
+		if ( empty( $token_id ) ) {
+			return [ 'valid' => false ];
+		}
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			$token = Keyring::init()->get_token_store()->get_token( array( 'type' => 'access', 'id' => $token_id ) );
+
+			return [
+				'valid' => ! empty( $token ),
+				'legacy' => $token && 'instagram' === $token->name,
+			];
+		}
+
+		$site = Jetpack_Options::get_option( 'id' );
+		$path = sprintf( '/sites/%s/instagram/%s/check-token', $site, $token_id );
+		$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+		$response_code = wp_remote_retrieve_response_code( $result );
+		if ( 200 !== $response_code ) {
+			return [
+				//We assume the token is valid if the response_code is anything but the invalid
+				//token codes we send back. This is to make sure it's not reset, if the API is down
+				//or something
+				'valid' => ! ( 403 == $response_code || 401 === $response_code ),
+				'legacy' => 'ERROR',
+			];
+		}
+		$status = json_decode( $result['body'], true );
+		return $status;
+	}
+
+	/**
+	 * Validates the widget instance's token ID and then uses it to fetch images from Instagram.
+	 * It then caches the result which it will use on subsequent pageviews.
+	 * Keyring is not loaded nor is a remote request is not made in the event of a cache hit.
+	 *
+	 * @param array $instance A widget $instance, as passed to a widget's widget() method.
+	 * @return string|array A string on error, an array of images on success.
+	 */
+	public function get_data( $instance ) {
+		if ( empty( $instance['token_id'] ) ) {
+			return 'ERROR';
```

Comment:
Would it be good to take this opportunity to refactor this to return `WP_Error` objects, rather than a generic `'ERROR'` string? Perhaps that's a separate PR.